pkg/lazy/lazy.go

```diff
@@ -0,0 +1,78 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lazy
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// Lazy represents a value whose computation is deferred until the first access
+type Lazy[T any] interface {
+	// Get returns the value, computing it if necessary.
+	Get() (T, error)
+}
+
+type lazyImpl[T any] struct {
+	getter func() (T, error)
+	retry  bool
+
+	res T
+	err error
+
+	done uint32
+	m    sync.Mutex
+}
+
+var _ Lazy[any] = &lazyImpl[any]{}
+
+// New returns a new lazily computed value. The value is guaranteed to only be computed a single time.
+func New[T any](f func() (T, error)) Lazy[T] {
+	return &lazyImpl[T]{getter: f}
+}
+
+// NewWithRetry returns a new lazily computed value. The value will be computed on each call until a
+// non-nil error is returned.
+func NewWithRetry[T any](f func() (T, error)) Lazy[T] {
+	return &lazyImpl[T]{getter: f, retry: true}
+}
+
+func (l *lazyImpl[T]) Get() (T, error) {
+	if atomic.LoadUint32(&l.done) == 0 {
+		// Outlined slow-path to allow inlining of the fast-path.
+		return l.doSlow()
+	}
+	return l.res, l.err
+}
+
+func (l *lazyImpl[T]) doSlow() (T, error) {
+	l.m.Lock()
+	defer l.m.Unlock()
+	if l.done == 0 {
```

Comment:
I guess you're relying on the lock forcing a sync across all threads, so you don't have to do an atomic get?

Reply:
Explained here: https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/sync/once.go;l=49-53 . This code is mostly copy+pasted with minor tweaks to support our needs. I thought I had commented that but it seems not -- I updated the comments a bit